server/src/test/java/org/opensearch/index/engine/InternalEngineTests.java

```diff
@@ -7546,6 +7546,61 @@ public void testGetSegmentInfosSnapshot() throws IOException {
         engine.close();
     }
 
+    // This method simulates behaviour of on-disk SegmentInfos having higher segment generation number compared to
+    // actual in-memory SegmentInfos and verified that engine.getSegmentInfosSnapshot returns on-disk SegmentInfos
+    public void testGetSegmentInfosWithHighestGenOnDisk() throws IOException {
+        IOUtils.close(store, engine);
+        Store store = spy(createStore());
+        InternalEngine engine = createEngine(store, createTempDir());
+        final int numDocs = randomIntBetween(10, 100);
+        for (int docId = 0; docId < numDocs; docId++) {
+            index(engine, docId);
+            if (randomBoolean()) {
+                engine.refresh("test");
+            }
+        }
+        engine.flush(true, true);
+
+        SegmentInfos sisDisk = store.readLastCommittedSegmentsInfo();
+        // Increment generation number of on-disk SegmentInfos
+        sisDisk.setNextWriteGeneration(sisDisk.getGeneration() + 1);
+
+        when(store.getLastCommitGeneration()).thenReturn(sisDisk.getGeneration());
+        when(store.readLastCommittedSegmentsInfo()).thenReturn(sisDisk);
+
+        GatedCloseable<SegmentInfos> segmentInfosSnapshot = engine.getSegmentInfosSnapshot();
+        assertEquals(segmentInfosSnapshot.get(), sisDisk);
+        assertEquals(segmentInfosSnapshot.get().getGeneration(), sisDisk.getGeneration());
+        segmentInfosSnapshot.close();
+        store.close();
+        engine.close();
+    }
+
+    // This method simulates behaviour of on-disk SegmentInfos having lower segment generation number compared to
+    // actual in-memory SegmentInfos and verified that engine.getSegmentInfosSnapshot returns in-memory SegmentInfos
+    public void testGetSegmentInfosWithHighestGenInMemory() throws IOException {
```

Comment:
-> This test case will have both in-memory and on-disk  segmentInfos  on same segment generation number but not lower on on-disk as it is mentioned (or purpose of this test case). In this test case we are not incrementing or decrementing segment generation of either on-disk on in-memory, so both will be same.  So we are not testing the case here that we actually want to.

-> Is this test really necessary? Will there be a scenario ever with on-disk segmentInfos having lower segment generation number than in-memory ?

Reply:
1. Looks like comment is outdated. Updated. 
2. Yes, this is useful to ensure in memory copy gets picked up (when segment generation is higher/equal compared to on-disk copy's segment generation). Yes, this scenario is happening during replica promotion.